docs/references/testing/page.md

```diff
@@ -0,0 +1,199 @@
+	Title string `json:"title"`
+}
+
+func Add(ctx *gofr.Context) (interface{}, error) {
+	var book Book
+
+	if err := ctx.Bind(&book); err != nil {
+		ctx.Logger.Errorf("error in binding: %v", err)
+		return nil, http.ErrorInvalidParam{Params: []string{"body"}}
+	}
+
+	// we assume the `id` column in the database is set to auto-increment.
+	res, err := ctx.SQL.ExecContext(ctx, `INSERT INTO books (title, isbn) VALUES (?, ?)`, book.Title, book.ISBN)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := res.LastInsertId()
+	if err != nil {
+		return nil, err
+	}
+
+	return id, nil
+}
+
+func main() {
+	// initialise gofr object
+	app := gofr.New()
+
+	app.POST("/book", Add)
+
+	// Run the application
+	app.Run()
+}
+
+```
+
+Here’s how to write tests using GoFr:
+
+```go
+// main_test.go
+package main
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"errors"
+	"net/http"
+	"testing"
+
+	"github.com/DATA-DOG/go-sqlmock"
+	"github.com/stretchr/testify/assert"
+
+	"gofr.dev/pkg/gofr"
+	"gofr.dev/pkg/gofr/container"
+	gofrHttp "gofr.dev/pkg/gofr/http"
+)
+
+func TestAdd(t *testing.T) {
+	type res struct {
+		ISBN interface{}
+		err  error
+	}
+
+	// NewMockContainer provides mock implementations for various databases including:
+	// Redis, SQL, ClickHouse, Cassandra, MongoDB, and KVStore.
+	// These mock can be used to define database expectations in unit tests,
+	// similar to the SQL example demonstrated here.
+	c, mock := container.NewMockContainer(t)
+
+	ctx := &gofr.Context{
+		Context:   context.Background(),
+		Request:   nil,
+		Container: c,
+	}
+
+	tests := []struct {
+		name        string
+		isBindError bool
+		mockExpect  func()
+		expectedRes interface{}
+	}{
+		{
+			name: "Error while Binding",
+			isBindError: true,
+			mockExpect: func() {
+			},
+			expectedRes: res{
+				nil,
+				gofrHttp.ErrorInvalidParam{Params: []string{"body"}}},
+		},
+		{
+			name: "Successful Insertion",
+			mockExpect: func() {
+				mock.SQL.
+					EXPECT().
+					ExecContext(ctx, `INSERT INTO books (title, isbn) VALUES (?, ?)`, "Book Title", 12345).
+					Return(sqlmock.NewResult(12, 1), nil)
+			},
+			expectedRes: res{
+				int64(12),
+				nil,
+			},
+		},
+		{
+			name: "Error on Insertion",
+			mockExpect: func() {
+				mock.SQL.
+					EXPECT().
+					ExecContext(ctx, `INSERT INTO books (title, isbn) VALUES (?, ?)`, "Book Title", 12345).
+					Return(nil, sql.ErrConnDone)
+			},
+			expectedRes: res{
+				nil,
+				sql.ErrConnDone},
+		},
+		{
+			name: "Error while fetching LastInsertId",
+			mockExpect: func() {
+				mock.SQL.
+					EXPECT().
+					ExecContext(ctx, `INSERT INTO books (title, isbn) VALUES (?, ?)`, "Book Title", 12345).
+					Return(sqlmock.NewErrorResult(errors.New("mocked result error")), nil)
+			},
+			expectedRes: res{
+				nil,
+				errors.New("mocked result error")},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tt.mockExpect()
+
+			var req *http.Request
+
+			if tt.isBindError {
```

Comment:
it is not advisable to have `if` condition in tests.. instead of `isBindError` we can use `body` field in `tests` struct, and give the respective body in testcases, and use the same while creating a request.